### DISCUSSION TITLE
Coupling (gradients of) vectors of variables in MOOSE

### DISCUSSION THREAD

**Discussion by richmondodufisan** [idaholab/moose#25907] [Q&A General]
The documentation (https://mooseframework.inl.gov/source/interfaces/Coupleable.html) is not fully clear on this. 

I have the following variables defined in my input file:

```
[GlobalParams]
  q_trans = 'q_trans_x q_trans_y q_trans_z' 
  order = FIRST
  family = LAGRANGE
[]

[Variables]
  [q_trans_x]
  []
  
  [q_trans_y]
  []
  
  [q_trans_z]
  []
[]
```

Defined with the GlobalParams block because I intend to use that vector of variables in several locations. I also have some places where I need to use the individual components alone. The kernel has:

```
  [diffuse_x_trans]
    type = DiffusionFlux
    variable = q_trans_x
    component_flux = 0
    flux_vector = q_trans
    block = transducer_material
  []
```

 Now for the source and header- following the syntax in the documentation, I have this declaration:

```
    unsigned int _n_vars;
    std::vector<MooseVariable *> _flux_vars;
```

And this implementation:

`  params.addRequiredCoupledVar("flux_vector", "Coupled value");`

```
	// Calculates number of components in vector & adds them to C++ vector
	_n_vars(coupledComponents("flux_vector"))
{
    for (unsigned int i = 0; i < _n_vars; i++)
      _flux_vars.push_back(dynamic_cast<MooseVariable *>(getVar("v", i)));
}
```

What I don't understand is this line: "_...Then, individual components can be obtained by calling coupledValue (or any other method mentioned above) passing in the variable name (as usual) and the component index..._"

Does the `_flux_vars.push_back(dynamic_cast<MooseVariable *>(getVar("v", i)));` not save the value of each of the coupled variables in that vector? Where do I call `coupledValue`? In my case, I want their gradients- where/how do I call `coupledGradients`? Should I be replacing `getVar` with something else?

And I understand that all MOOSE variables need to be accessed at the qudrature point being solved. Is the correct syntax `_flux_vars[_qp](j)` for `j=1` to `_n_vars`? Again, how do I get the gradient of each variable?

**Comment by GiudGiud** [comment]
Hello

You can look at CoupledForce for an example but essentially, coupledValue and getVar are completely separate APIs.
getVar is used to get a pointer to the variable, for various purposes but if you need the value you can use (among others) the "functor interfaces". 

so (*var)(ElementQpArg (see MooseFunctorArguments.h)) 
EDIT: functors not available for array variables!

coupledValue is must simpler to use and gets you a reference to the current value of the variable on the Qps of the element. It also handles array variables if that's what you have right now with coupledArrayValue(s).

You use it by retrieving the reference in the constructor initializer's list. For example here for 1 component (you ll want to adapt this to retrieve all components, see Coupleable.h)
In header:

const ArrayVariableValue & var_comp_0;

in constructor initializer list
```
...,
var_comp_0(coupledArrayValue("var_name_in_parameter", 0),
...
```

**Reply by richmondodufisan** [reply]
Thank you, that makes sense. Is it possible to define a vector `std::vector<ArrayVariableGradient> _flux_vars;` so that I can use coupledComponents to find the number of components the user provides, and then do `_flux_vars.push_back(coupledArrayGradient("var_name_in_parameter", i);`, where i ranges from 0 to the number of components within a for loop? Or do I need to do them all individually?

**Reply by GiudGiud** [reply]
So the difficulty with that is that C++ does not support vectors of references (among so many other things, I m not a fan of C++ anymore). And you MUST use references otherwise the (array gradient) value does not get updated

So you ll need to do that individually, on a per-variable component basis. 

Unless there is a coupledArrayGradients that gets you a tensor with all the gradients. Did you check the coupleable interface?

**Reply by richmondodufisan** [reply]
Ah I see. Yeah I checked coupleable and it seems like working with the individual components will have to do for now. Thanks! I'll check back here if I run into issues with that.

**Reply by richmondodufisan** [reply]
@GiudGiud does `var_comp_0(coupledArrayValue("var_name_in_parameter", 0),` retrieve the 1st component in the input file, or does it retrieve the whole vector? Asking because I got an error saying `var_comp_0` was a matrix i.e `'const Eigen::Matrix<double, -1, 1>'` that went away after I indexed `var_comp_0(i)`

**Reply by GiudGiud** [reply]
Seems the zero you passed is the index in the vector of array variables that you could have passed to the parameter 

So it s the whole variable

**Reply by GiudGiud** [reply]
CoupledArrayGradient probably behaves the same way you should be able to get all the gradients at once?

**Reply by richmondodufisan** [reply]
So I tried with coupledArrayGradient and it ended up giving me a 'const Eigen::Matrix<double, -1, 3>', which seemed to suggest that for the coupledArrayValue, it really is just a single value, even though its storing it in a Matrix? Or what do you think?

Then I also ran into another problem- I took out the GlobalParameters block and instead define the vector of variables individually at each location I need them. That is, within the relevant blocks,

`v = 'q_trans_x q_trans_y q_trans_z'`

But I'm getting the error: `Variable 'sample_flux_vector' is of a different C++ type than you tried to fetch it as.`

When I left in the GlobalParams block, with

```
[GlobalParams]
  q_trans = 'q_trans_x q_trans_y q_trans_z' 
[]
```

and then 

`v = q_trans`

It said `Coupled variable 'q_samp' was not found`

How am I supposed to define the vector of variables in the input file?

**Reply by richmondodufisan** [reply]
I'm suspecting that it has to do with `params.addRequiredCoupledVar("v", "the flux vector of the transducer");` , but there is no `params.addRequiredCoupledArray`

edit: I ended up just coupling each individual variable/gradient I need instead of trying to use an array

**Reply by GiudGiud** [reply]
addRequiredCoupledVar should work the same way for all types of variables. It s the "get" or coupeldXX routines that are differenct. 

`Coupled variable 'q_samp' was not found` is an odd message for the text you have pasted as it does not refer to q_samp anywhere?
Look for a parameter where you passed q_samp instead of q_trans

**Reply by richmondodufisan** [reply]
Oh yeah I just edited the variable name to be consistent here, just forgot to edit the error lol. But in the input file I actually had both q_samp and q_trans and had passed q_samp earlier so it saw that one first. Using individual coupled variables as before worked though.
